### ISSUE TITLE
Regression: Port forwarding fails while working with remote SSH

### ISSUE THREAD

**Issue by Ashesh3** [microsoft/vscode#192521]
This is the 4th time this issue has occurred and fixed, then regressed within the last 4 months.. Such frequent major bugs make VScode really unstable to use :/

Does this issue occur when all extensions are disabled?: Yes

```
Version: 1.82.0 (user setup)
Commit: 8b617bd08fd9e3fc94d14adb8d358b56e3f72314
Date: 2023-09-06T22:07:07.438Z
Electron: 25.8.0
ElectronBuildId: 23503258
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Windows_NT x64 10.0.19045
```

Steps to Reproduce: (See https://github.com/microsoft/vscode/issues/192521#issuecomment-1714483320 )

1. Connect to a remote ssh host via remote explorer
2. Start a server on 0.0.0.0:3000
3. Port forward it
4. Try to access it on the current PC.

Error logs in `Shared` window: (Could be irrelevant since I get the log below on both 1.81.2 and 1.81.1 but no issue is faced on 1.18.1)
```
2023-09-08 12:03:50.235 [error] [uncaught exception in sharedProcess]: Cannot read properties of undefined (reading 'then'): TypeError: Cannot read properties of undefined (reading 'then')
    at v.s (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:26:4731)
    at v.q (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:26:4161)
    at c.value (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:26:3566)
    at b.w (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:21:1902)
    at b.x (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:21:1972)
    at b.fire (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:21:2188)
    at MessagePortMain.te (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:19:40559)
    at MessagePortMain.emit (node:events:513:28)
    at MessagePortMain._internalPort.emit (node:electron/js2c/utility_init:2:367)
2023-09-08 12:03:51.522 [error] [uncaught exception in sharedProcess]: Cannot read properties of undefined (reading 'then'): TypeError: Cannot read properties of undefined (reading 'then')
    at v.s (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:26:4731)
    at v.q (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:26:4161)
    at c.value (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:26:3566)
    at b.w (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:21:1902)
    at b.x (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:21:1972)
    at b.fire (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:21:2188)
    at MessagePortMain.te (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:19:40559)
    at MessagePortMain.emit (node:events:513:28)
    at MessagePortMain._internalPort.emit (node:electron/js2c/utility_init:2:367)
2023-09-08 12:03:52.335 [error] [uncaught exception in sharedProcess]: Cannot read properties of undefined (reading 'then'): TypeError: Cannot read properties of undefined (reading 'then')
    at v.s (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:26:4731)
    at v.q (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:26:4161)
    at c.value (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:26:3566)
    at b.w (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:21:1902)
    at b.x (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:21:1972)
    at b.fire (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:21:2188)
    at MessagePortMain.te (c:\Users\Ashesh\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\code\node\sharedProcess\sharedProcessMain.js:19:40559)
    at MessagePortMain.emit (node:events:513:28)
    at MessagePortMain._internalPort.emit (node:electron/js2c/utility_init:2:367)
```

launch.json:

```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Next.js: debug server-side",
            "type": "node-terminal",
            "request": "launch",
            "command": "npm run dev"
        },
        {
            "name": "Next.js: debug client-side",
            "type": "chrome",
            "request": "launch",
            "url": "http://localhost:3000"
        },
        {
            "name": "Next.js: debug full stack",
            "type": "node-terminal",
            "request": "launch",
            "command": "npm run dev",
            "serverReadyAction": {
                "pattern": "started server on .+, url: (https?://.+)",
                "uriFormat": "%s",
                "action": "debugWithChrome"
            },
            "sourceMaps": true,
        }
    ]
}
```

![image](https://github.com/microsoft/vscode/assets/3626859/bab54f52-4074-42e2-8f49-18a16c8460b6)

**Comment by Ashesh3** [comment]
Everything works fine on the previous version:
```
Version: 1.81.1 (user setup)
Commit: 6c3e3dba23e8fadc360aed75ce363ba185c49794
Date: 2023-08-09T22:22:42.175Z
Electron: 22.3.18
ElectronBuildId: 22689846
Chromium: 108.0.5359.215
Node.js: 16.17.1
V8: 10.8.168.25-electron.0
OS: Windows_NT x64 10.0.19045
```

**Comment by ilteoood** [comment]
I don't know if this can help, but on Linux this is working
![image](https://github.com/microsoft/vscode/assets/6383527/aad5d08a-4545-44ed-b1e3-41b35b225b75)

**Comment by niranjan94** [comment]
Can confirm. Facing the same issue on Mac (Client), Fedora (Remote).

```
Version: 1.82.0
Commit: 8b617bd08fd9e3fc94d14adb8d358b56e3f72314
Date: 2023-09-06T22:09:41.364Z
Electron: 25.8.0
ElectronBuildId: 23503258
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Darwin x64 21.6.0
```

**Comment by WANGBohaO-jpg** [comment]
I meet the same problem!

**Comment by gustalima** [comment]
Same issue here. 

Forwarding the ports from the terminal works, though.

**Comment by misl-smlz** [comment]
We are facing the same issue (vs code under windows with remote ssh against ubuntu). Rolling back to https://code.visualstudio.com/updates/v1_81 fixes the issue.

**Comment by au2001** [comment]
Maybe related to #191186

Facing the same issue from an Apple silicon client (`Darwin arm64 22.6.0`) to an Ubuntu 20.04 server

**Comment by rkulow** [comment]
same issue here:

```
Version: 1.82.0 (user setup)
Commit: 8b617bd08fd9e3fc94d14adb8d358b56e3f72314
Date: 2023-09-06T22:07:07.438Z
Electron: 25.8.0
ElectronBuildId: 23503258
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Windows_NT x64 10.0.22621

```

**Comment by jasonwashburn** [comment]
Experiencing same with Windows client via SSH to RH7 host. Prior to update, worked without issue...now we get a lot of weird/slow/unusable connections when trying to forward traffic from host port to localhost port for testing.

**Comment by anastasiuspernat** [comment]
Same issue on macOS Ventura. Remote-SSH starts to freeze because of that.

**Comment by chanbyeoldaddy** [comment]
@Ashesh3 
Would you share how to rollback to 6c3e3dba23e8fadc360aed75ce363ba185c49794 on window machine? (window 10)?

**Comment by Ashesh3** [comment]
> @Ashesh3 Would you share how to rollback to [6c3e3db](https://github.com/microsoft/vscode/commit/6c3e3dba23e8fadc360aed75ce363ba185c49794) on window machine? (window 10)?

Look for download option just below the header on this page:
https://code.visualstudio.com/updates/v1_81

For windows:
https://update.code.visualstudio.com/1.81.1/win32-x64-user/stable

**Comment by yueyericardo** [comment]
Same happened to me on mac Ventura (client) and ubuntu (Remote)

**Comment by alexr00** [comment]
I'm not able to reproduce this issue. My steps:

1. Connect to a remote with the Remote SSH extension
2. Start a process on the remote that listens on `0.0.0.0:3000` (I also tried `localhost:3000`)
3. The port gets auto forwarded
4. Open in browser. I tried the following ways to open:
  - ctrl+click on the link in the terminal
  - Use the globe icon in the ports vie
5. See my expected page in the browser.

I tested local Windows -> remote Ubuntu and local Windows -> remote Windows.

I specifically used https://github.com/microsoft/vscode-remote-try-node as my test. Is anyone in this thread able to reproduce the issue with that repo?

**Comment by au2001** [comment]
@alexr00 I have only been able to reproduce it with real-word projects for now.
I believe the difference lies in the size transferred, as requests start to hang only after the first few 100s of KB (or few MBs?) are transferred from the server.
I can try to create a minimal reproducible example based on the repo you linked when I get back from work in a few hours.

**Comment by alexr00** [comment]
I think I have a repro:

1. Start with https://github.com/microsoft/vscode-remote-try-node
2. Add a lot more text to the `res.send`
3. `npm run start`
4. Open the page in the browser
5. kill the process form step 3
6. `npm run start` again
7. Open the page in the browser. It will spin indefinitely.

**Comment by connor4312** [comment]
Thanks @alexr00, I was able to repro a problem with those steps. THat looks the same as 

However @ashesh3 it's hard for me to tell just based on"Try to access it on the current PC" whether this is the same problem you're hitting. Can you describe your problem in more details, or grab a screencap of what you hit? https://gifcap.dev/ is useful

Note that we have rolled back the change that probably caused this issue, so to reproduce it again you may need to add `"remote.SSH.useExecServer": true` to your user settings. Thanks!

**Comment by au2001** [comment]
> @alexr00 I think I have a repro

Can confirm this reproduces the exact problem I was having.
Using `npx nodemon server.js` at step 3 and updating the `server.js` file instead of steps 5 & 6 also works, which is probably the most common scenario for this bug.

> @connor4312 Note that we have rolled back the change that probably caused this issue, so to reproduce it again you may need to add `"remote.SSH.useExecServer": true` to your user settings. Thanks!

Don't know about @Ashesh3 & others, but with `"remote.SSH.useExecServer": false` in my user settings, all issues are fixed on my side. Thanks for this tip!

**Comment by Ashesh3** [comment]
> Thanks @alexr00, I was able to repro a problem with those steps. THat looks the same as #192627
> 
> However @Ashesh3 it's hard for me to tell just based on "Try to access it on the current PC" whether this is the same problem you're hitting. Can you describe your problem in more detail, or grab a screencap of what you hit? https://gifcap.dev/ is useful
> 
> Note that we have rolled back the change that probably caused this issue, so to reproduce it again you may need to add `"remote.SSH.useExecServer": true` to your user settings. Thanks!

@connor4312  Hey, apologies for the non-descriptive steps to replicate the issue. What @alexr00 mentioned above is exactly the steps that pertains to the issue I have and the case which @au2001 mentioned above is specifically relevant to me. Running the server node process, then changing the contents of the server file followed by restarting the server (it happens automatically in my case) causes the browser to get stuck in an infinite loading state.


(Note: The below is reported with `"remote.SSH.useExecServer": true` as per your suggestion)

On my windows vscode client (connected to macOS host):
- Start vite server
- Make a file change while the server is running
- Terminate the process (ctrl+c)
- Start the vite server again
- Now, the browser will not serve any requests from the same port even after vite has started completely and is ready to serve
- It basically hangs in a loading state

![Recording 2023-09-12 at 01 18 17](https://github.com/microsoft/vscode/assets/3626859/c48d1ec1-633a-4350-97cf-8dd4d633fde9)



Setting `"remote.SSH.useExecServer": false` fixes the above issue for me. Everything loads up instant no matter how many times and when i restart my vite server.

**Comment by connor4312** [comment]
Excellent, thanks for confirming. I will duplicate the other issue into this one.

**Comment by connor4312** [comment]
@alexr00 I'm downgrading this from a candidate since we backed out the SSH change that exposed this issue.

**Comment by Ashesh3** [comment]
@connor4312 Your linked PR closed this issue even though the PR says, "Part 1 of fixing", just checking if it was closed accidently.

**Comment by LeonardoEmili** [comment]
Probably unnecessary, but reporting here in case any user comes across this issue. For me, vscode automatically updated to `1.81.2` and I was getting the following error, nothing worked (removing the local nor remote `.vscode` directory, restarting the server, reinstalling vscode) but I had to manually downgrade to `1.81.1`. The error I was getting below:
```
[10:19:08.867] Failed to parse remote port from server output
[10:19:08.868] Resolver error: Error: 
	at g.Create (/Users/leonardoemili/.vscode/extensions/ms-vscode-remote.remote-ssh-0.106.3/out/extension.js:2:637876)
	at t.handleInstallOutput (/Users/leonardoemili/.vscode/extensions/ms-vscode-remote.remote-ssh-0.106.3/out/extension.js:2:635279)
	at t.tryInstall (/Users/leonardoemili/.vscode/extensions/ms-vscode-remote.remote-ssh-0.106.3/out/extension.js:2:756875)
	at async /Users/leonardoemili/.vscode/extensions/ms-vscode-remote.remote-ssh-0.106.3/out/extension.js:2:717467
	at async t.withShowDetailsEvent (/Users/leonardoemili/.vscode/extensions/ms-vscode-remote.remote-ssh-0.106.3/out/extension.js:2:720720)
	at async C (/Users/leonardoemili/.vscode/extensions/ms-vscode-remote.remote-ssh-0.106.3/out/extension.js:2:714381)
	at async t.resolve (/Users/leonardoemili/.vscode/extensions/ms-vscode-remote.remote-ssh-0.106.3/out/extension.js:2:718116)
	at async /Users/leonardoemili/.vscode/extensions/ms-vscode-remote.remote-ssh-0.106.3/out/extension.js:2:901786
[10:19:08.869] ------
```
@connor4312 I understand that you rolled back to the previous SSH behaviour since version `1.81.1` has been removed from the releases section.

**Comment by connor4312** [comment]
@Ashesh3 the other parts were in the closed-source ssh extension

@LeonardoEmili your issue is unrelated, please open a new issue and share the complete Remote SSH output
